src/transform/src/reduce_reduction.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Breaks complex `Reduce` variants into a join of simpler variants.
+//!
+//! Specifically, any `Reduce` that contains two different "types" of aggregation,
+//! in the sense of `ReductionType`, will be broken in to one `Reduce` for each
+//! type of aggregation, each containing the aggregations of that type,
+//! and the results are then joined back together.
+
+use crate::TransformCtx;
+use mz_compute_types::plan::reduce::reduction_type;
+use mz_expr::MirRelationExpr;
+
+/// Breaks complex `Reduce` variants into a join of simpler variants.
+#[derive(Debug)]
+pub struct ReduceReduction;
+
+impl crate::Transform for ReduceReduction {
+    /// Transforms an expression through accumulated knowledge.
+    #[mz_ore::instrument(
+        target = "optimizer",
+        level = "debug",
+        fields(path.segment = "reduce_reduction")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: &mut TransformCtx,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_pre_mut(&mut Self::action);
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl ReduceReduction {
+    /// Breaks complex `Reduce` variants into a join of simpler variants.
+    pub fn action(relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic,
+            expected_group_size,
+        } = relation
+        {
+            // Segment the aggregates by reduction type.
+            let mut segmented_aggregates = std::collections::BTreeMap::default();
+            for (index, aggr) in aggregates.iter().enumerate() {
+                let (aggrs, indexes) = segmented_aggregates
+                    .entry(reduction_type(&aggr.func))
+                    .or_insert_with(|| (Vec::default(), Vec::default()));
+                indexes.push(group_key.len() + index);
+                aggrs.push(aggr.clone());
+            }
+
+            // Do nothing unless there are at least two distinct types of aggregations.
+            if segmented_aggregates.len() < 2 {
+                return;
+            }
+
+            // For each type of aggregation we'll plan the corresponding `Reduce`,
+            // and then join the at-least-two `Reduce` stages together.
+            // TODO: Perhaps we should introduce a `Let` stage rather than clone the input?
```

Comment:
Re. your TODO: What do we do elsewhere? It seems better to introduce a `Let` stage rather than computing the same result again, or can we rely on subsequent optimization passes to clean this up for us?

Reply:
`RelationCSE` is run after this pretty soon in the optimizer pipeline, and it should take care of this. Looking at the plans, it indeed does seem to take care of this.